### DISCUSSION TITLE
M5stack III Env sensor QMP6988 Support

### DISCUSSION THREAD

**Discussion by ghost** [arendst/Tasmota#13833] [Ideas]
Hello,

I hope this is the right place for my question. I've recently bought a M5stack III Env sensor, which has a QMP6988 at I2C address 0x70. The SHT3x is already working fine, but the Barometric pressure not yet.

https://www.tinytronics.nl/shop/nl/sensoren/lucht/druk/m5stack-env-iii-unit

I2Cscan shows a detected device at address 0x70. I'm running the tasmota-sensors release.

Would it be possible to add support for the Barometric pressure for this sensor? Or is there a way I can add it myself?

Thanks a bunch in advance!

**Comment by sfromis** [comment]
Note that the tasmota-sensors binary by far does not have all sensors, just a selection. However, the QPM6988 does not seem to be among the optional ones, unless it "happens" to be compatible with some other sensor.

Support for new sensors is indeed possible, if "someone" creates a driver. The project welcomes a PR for additional sensor support. With ESP32, you also have the option of creating a driver in the Berry language, if you are not a fan of programming in C++.

**Comment by Slyke** [comment]
Hey @WhizzCat what did you do to get the temp sensor working? Can't seem to get Tasmota to detect it and show the values on the webpage.

**Reply by ghost** [reply]
Nothing really, I just added it to configured I2C ports. Temp, humidity and DewPoint work fine, just not the barometric pressure which is supposed to be supported. I haven't tried on an Arduino or something similar. Maybe you got another device on the I2C bus with a conflicting address? Could be worth to put it solo on two other pins as I2C.

![image](https://user-images.githubusercontent.com/17166886/148679800-31eaeaab-2ec9-4e0c-bf53-4ed350bb3a30.png)

**Reply by sfromis** [reply]
Does that mean that QMP6988 is compatible with SHT3X?

**Reply by Slyke** [reply]
Hmmm, not sure. I compiled Tasmota.core2 and flashed it onto the m5Stack. I see the following on the main page:
![image](https://user-images.githubusercontent.com/6442613/148680540-ae5aeb6b-8da3-454a-b053-e97ecfb4a639.png)

I've plugged the temp sensor in and verified it does work by using the example code provided with the ENV III library, but Tasmota doesn't seem to want to detect it. In the templates section I have GPIO32 set to SDA1 and GPIO33 set to SCL1:
![image](https://user-images.githubusercontent.com/6442613/148680593-0e6d22a7-1c22-4bc2-abef-47ba7480a8ec.png)

I also don't have any of those icons showing.

This is the template I used:
```
{"NAME":"M5Stack Core 2","GPIO":[6210,1,1,1,6720,768,0,0,0,1,1,800,0,0,736,1,0,641,609,704,0,0,0,1,0,0,0,0,640,608,1,1,1,0,672,0],"FLAG":0,"BASE":1}
```

It seems it supports everything... except the external temperature sensor. Haven't tried another I2C device yet.

**Reply by ghost** [reply]
The screenshot is from my Home Assistant, so that counts for the icons, it was just to show it works.

Anyway, I've flashed Tasmota Sensors on a Wemos D1 Mini (ESP8266) and it works out of the box. I've combined it withm several other I2C sensors. I do believe this thing actually routes the I2C bus internally, because it gives off two I2C addresses.

Can you do an i2cscan on the console of your Tasmota device? That way you should be able to see at least if it gets detected properly on the device. I think it's compatible with the SHT though. The values make sense compared to my other sensors :)

**Reply by barbudor** [reply]
> Does that mean that QMP6988 is compatible with SHT3X?

@sfromis no
The [env III sensor](https://shop.m5stack.com/collections/m5-sensor/products/env-iii-unit-with-temperature-humidity-air-pressure-sensor-sht30-qmp6988) has both a SHT3X (temp+hum) and a QMP6988 (pressure).
So still only the SHT is working until someone open VSCode and write a driver for QMP6988...

**Reply by sfromis** [reply]
Yeah, the need for a driver was my initial assumption, and my question was largely rhetoric, to highlight that SHT3X working was basically off-topic.

**Reply by Slyke** [reply]
> Can you do an i2cscan on the console of your Tasmota device?

```
22:25:47.076 CMD: I2Cscan
22:25:47.096 RSL: RESULT = {"I2CScan":"Device(s) found at 0x44 0x70"}
```

I believe the temperature and humidity sensor is 0x44 and the QMP6988 sensor is 0x70, according to this:
https://shop.m5stack.com/products/env-iii-unit-with-temperature-humidity-air-pressure-sensor-sht30-qmp6988

So Tasmota can see them, but not sure how to get it to read the values from them. It should just do it automatically like when connecting most I2C devices to it? It might be something specifically with the `tasmota32-core2` build.

**Reply by sfromis** [reply]
The cause of not getting QMP6988 values is very simple, no one has submitted a Tasmota driver for this chip. Until there is a driver, Tasmota does not do anything with the chip. A driver can be done as a "native" driver coded in C++, or with Berry (when it is on ESP32), where coding I2C drivers can be done in a reasonably friendly environment, without the many pitfalls.
